Remove Stop and use Duration everywhere

simonrozsival opened the pull request:
Closes #747, Ref #709

Please look especially at the Ultrawave model - I use two fields for the duration property: `Duration` and `ApiDuration`. One of them is ignored for serialization. I decided to do it this way because it makes it easy to make sure that the internal `Duration` is `null` when it is running and when serialized the negative start timestamp is used and similarly for deserialization.

~~WIP: I still have to go through the code once more and make sure all of the `.Stop` usages are gone + do a functional test in the simulator.~~

Functional test shows that it works well after the refactoring (pushing and pulling data, duration edit view, ...).

There should be a follow-up issue to refactor the edit duration view, which still uses the stop time and then converts it to duration which is then passed to the entity. This PR is already quite big, I didn't want to include that in this PR which would make it a lot bigger.

Squash whenever.

amulware commented:
@simonrozsival since you know best what still has to be changed to remove all the `stop`s, can you create the followup issue?

amulware commented:
@arturs105 @kornelijepetak +1 review as sanity check please?

simonrozsival commented:
I already did - #810. Should I add more details to the issue or is it OK like this?

amulware commented:
Didn't see that, looks good!

kornelijepetak commented on Toggl.Ultrawave/Models/TimeEntry.cs:
```diff
@@ -20,14 +22,22 @@ internal sealed partial class TimeEntry : ITimeEntry
         {
-            get => Stop.HasValue ? (int)(Stop.Value - Start).TotalSeconds : -Start.ToUnixTimeSeconds();
+            get => Duration.HasValue ? Start.AddSeconds(Duration.Value) : null as DateTimeOffset?;
             set { }
```
Why do we allow an empty setter? I'm confused at the moment. It seems that there are lots of call sites that set the property to all sorts of values. But if it's just swallowed, what's it use? Or did I miss something?

simonrozsival replied:
This is because of deserialization - it needs a setter for that value to work, but we don't need the value, so we simply ignore it.

kornelijepetak replied:
That's what I thought. However, that does not explain why there are other callsites (except the serialization) where this property gets assigned a specific value.

simonrozsival replied:
There should be none. This property is not in the `Toggl.Multivac.Models.ITimeEntry`, it is just in `Toggl.Ultrawave.Models.TimeEntry`. I found some leftovers in the tests, I will :fire: these.

kornelijepetak replied:
True, those were only in tests.

simonrozsival replied:
Thanks for pointing this out!

simonrozsival commented on Toggl.Foundation.Tests/Sync/ConflictResolution/TimeEntryRivalsResolverTests.cs:
```diff
@@ -106,25 +109,25 @@ public void TheStoppedTimeEntryMustBeMarkedAsSyncNeededAndTheStatusOfTheOtherOne
             var (fixedA, _) = resolver.FixRivals(a, b, timeEntries);
 
-            fixedA.Stop.Should().Be(now);
+            fixedA.Duration.Should().Be(5 * 24 * 3600);
```
Yes, that is a good idea, I will change that.

amulware replied:
:+1:

simonrozsival commented:
I addressed all comments, there is no review to dismiss, there were no drastic changes, so I take it that your reviews still hold, when the tests run successfully, I will squash this.

amulware commented:
:shipit:
